sway/tree/workspace.c

```diff
@@ -817,3 +833,10 @@ size_t workspace_num_tiling_views(struct sway_workspace *ws) {
 	workspace_for_each_container(ws, count_tiling_views, &count);
 	return count;
 }
+
+void workspace_squash(struct sway_workspace *workspace) {
+	for (int i = 0; i < workspace->tiling->length; i++) {
+		struct sway_container *child = workspace->tiling->items[i];
```

Comment:
Can this access out of bounds memory if `workspace->tiling` shrinks in length during `container_squash`?

Reply:
Separately, I'm not really sure whether we should be relying on the order containers appear in `workspace->tiling` here (for the skipping). It seems fragile if we don't already do this elsewhere.

Reply:
I don't think workspace->tiling can shrink in length during the squash. container_squash only modifies the parent by adding containers to it, which is why I keep track of the number added. Containers added to the parent are not themselves eligible to be squashed because the recursion ends when it's done adding them and destroys the container pair. Basically each child has the opportunity to squash one pair of containers in its family, but it stops there.

Regardless, unless I'm sorely mistaken about C (possible), the condition statement is evaluated on each loop, so if length changes the condition value will reflect that and the loop will end rather then set `i >= workspace->length->tiling`. I expect that the length will grow because I am adding containers to it, so I actually need that for correctness. Otherwise the loop will end early.

Actually there is a bug though...  the reap_empty call should decrement change before returning because at that point con and child are certainly empty and will be removed. Atm I guess it skips the immediate sibling of children it successfully squashed.

Reply:
You're right; I wasn't paying close enough attention.